File: epi_judge_cpp/apply_permutation.cc
#include <vector>

#include "test_framework/generic_test.h"
using std::vector;

void ApplyPermutation(vector<int> perm, vector<int>* A_ptr) {
  // TODO - you fill in here.
  auto &A {*A_ptr};

  for (auto i = 0; i < perm.size(); i++) {
    auto idx = i;
    int temp = A[idx];
    while(perm[idx] != idx) {
      std::swap(A[perm[idx]], temp);
      std::swap(idx, perm[idx]);
    }
  }
  return;
}
vector<int> ApplyPermutationWrapper(const vector<int>& perm, vector<int> A) {
  ApplyPermutation(perm, &A);
  return A;
}

int main(int argc, char* argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"perm", "A"};
  return GenericTestMain(args, "apply_permutation.cc", "apply_permutation.tsv",
                         &ApplyPermutationWrapper, DefaultComparator{},
                         param_names);
}

File: epi_judge_python/apply_permutation.py
from typing import List

from test_framework import generic_test


def apply_permutation(perm: List[int], A: List[int]) -> None:
    for i in range(len(perm)):
        next = i
        while perm[next] > 0:
            A[i], A[perm[next]] = A[perm[next]], A[i]
            temp = perm[next]
            perm[next] -= len(A)
            next = temp
    return


def apply_permutation_wrapper(perm, A):
    apply_permutation(perm, A)
    return A


if __name__ == "__main__":
    exit(
        generic_test.generic_test_main(
            "apply_permutation.py", "apply_permutation.tsv", apply_permutation_wrapper
        )
    )
